File: go/internal/daemon/routes/approvals.go
/**
 * Author: Deepankar Das
 */

package routes

import (
	"encoding/json"
	"net/http"

	"github.com/anthropics/enforcer/internal/approval"
	"github.com/anthropics/enforcer/internal/types"
)

// HandleGetPending returns all currently pending approval requests.
func HandleGetPending(svc *approval.ApprovalService) (int, interface{}) {
	pending := svc.GetPending()
	return http.StatusOK, map[string]interface{}{
		"approvals": pending,
		"count":     len(pending),
	}
}

// HandleResolveApproval resolves a pending approval with the given decision.
func HandleResolveApproval(approvalID string, body []byte, svc *approval.ApprovalService) (int, interface{}) {
	if approvalID == "" {
		return http.StatusBadRequest, map[string]string{
			"error":   "missing_approval_id",
			"message": "Approval ID is required in the URL path.",
		}
	}

	var decision types.ApprovalDecision
	if err := json.Unmarshal(body, &decision); err != nil {
		return http.StatusBadRequest, map[string]string{
			"error":   "invalid_json",
			"message": "Failed to parse approval decision: " + err.Error(),
		}
	}

	// Validate decision field.
	if decision.Decision != "approve" && decision.Decision != "deny" {
		return http.StatusBadRequest, map[string]string{
			"error":   "invalid_decision",
			"message": "Decision must be 'approve' or 'deny'.",
		}
	}

	if err := svc.ResolveApproval(approvalID, decision); err != nil {
		return http.StatusNotFound, map[string]string{
			"error":   "approval_not_found",
			"message": err.Error(),
		}
	}

	return http.StatusOK, map[string]interface{}{
		"approval_id": approvalID,
		"decision":    decision.Decision,
		"resolved":    true,
	}
}

// HandleGetApprovalStatus returns the status of a single approval by ID.
// The hook handler polls this endpoint while waiting for admin approval.
func HandleGetApprovalStatus(approvalID string, svc *approval.ApprovalService) (int, interface{}) {
	if approvalID == "" {
		return http.StatusBadRequest, map[string]string{
			"error":   "missing_approval_id",
			"message": "Approval ID is required in the URL path.",
		}
	}

	status := svc.GetApprovalStatus(approvalID)
	if status == nil {
		return http.StatusNotFound, map[string]string{
			"error":   "approval_not_found",
			"message": "No approval found with ID: " + approvalID,
		}
	}

	return http.StatusOK, status
}

// HandleGetApprovalMetrics returns aggregate approval service metrics.
func HandleGetApprovalMetrics(svc *approval.ApprovalService) (int, interface{}) {
	metrics := svc.GetMetrics()
	return http.StatusOK, metrics
}

File: src/daemon/routes/approvals.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Approval API Routes
 *
 * REST endpoints for the approval workflow:
 *   POST   /v1/approvals           — create approval request
 *   GET    /v1/approvals/:id       — get approval status + context
 *   POST   /v1/approvals/:id/resolve — approve or deny
 *   GET    /v1/approvals/pending   — list all pending approvals
 *
 * TDD Reference: Section 10.3
 */

import type { ApprovalService } from "../../approval/service.js";
import { requestBreakGlass } from "../../approval/break-glass.js";

export interface ApprovalRouteResult {
  status: number;
  body: unknown;
}

/**
 * GET /v1/approvals/pending
 */
export function handleGetPending(approvalService: ApprovalService): ApprovalRouteResult {
  const pending = approvalService.getPending();
  return {
    status: 200,
    body: { approvals: pending, count: pending.length },
  };
}

/**
 * GET /v1/approvals/:id
 */
export function handleGetApproval(
  approvalId: string,
  approvalService: ApprovalService,
): ApprovalRouteResult {
  const result = approvalService.getApproval(approvalId);
  if (!result) {
    return { status: 404, body: { error: `Approval ${approvalId} not found` } };
  }
  return {
    status: 200,
    body: {
      approval_id: result.approval.approval_id,
      status: result.status,
      context_bundle: result.approval.context_bundle,
      timeout_seconds: result.approval.timeout_seconds,
      timeout_behavior: result.approval.timeout_behavior,
      created_at: result.approval.created_at,
    },
  };
}

/**
 * POST /v1/approvals/:id/resolve
 *
 * Body: { decision: "approve" | "deny", approver_id: string, rationale?: string, scope?: object, is_break_glass?: boolean }
 */
export function handleResolveApproval(
  approvalId: string,
  rawBody: string,
  approvalService: ApprovalService,
): ApprovalRouteResult {
  let body: Record<string, unknown>;
  try {
    body = JSON.parse(rawBody);
  } catch {
    return { status: 400, body: { error: "Invalid JSON" } };
  }

  const decision = body.decision as string;
  const approverId = body.approver_id as string;

  if (!decision || !["approve", "deny"].includes(decision)) {
    return { status: 400, body: { error: "decision must be 'approve' or 'deny'" } };
  }
  if (!approverId) {
    return { status: 400, body: { error: "approver_id is required" } };
  }

  try {
    // Handle break-glass
    if (body.is_break_glass) {
      const rationale = body.rationale as string;
      if (!rationale || rationale.trim().length === 0) {
        return { status: 400, body: { error: "Break-glass requires a non-empty rationale" } };
      }
      const bgDecision = requestBreakGlass(approvalId, approverId, rationale);
      approvalService.resolveApproval(approvalId, bgDecision);
      return {
        status: 200,
        body: { approval_id: approvalId, decision: "approve", is_break_glass: true },
      };
    }

    // Normal resolution
    approvalService.resolveApproval(approvalId, {
      approval_id: approvalId,
      decision: decision as "approve" | "deny",
      approver_id: approverId,
      rationale: (body.rationale as string) || undefined,
      scope: body.scope as { type: "single" | "session" | "time_bounded"; pattern?: string; expiry?: string } | undefined,
      is_break_glass: false,
    });

    return {
      status: 200,
      body: { approval_id: approvalId, decision, approver_id: approverId },
    };
  } catch (err) {
    const message = err instanceof Error ? err.message : "Unknown error";
    return { status: 404, body: { error: message } };
  }
}

/**
 * GET /v1/approvals/metrics
 */
export function handleGetApprovalMetrics(approvalService: ApprovalService): ApprovalRouteResult {
  return {
    status: 200,
    body: approvalService.getMetrics(),
  };
}